engine/apps/integrations/mixins/alert_channel_defining_mixin.py

```diff
@@ -29,83 +31,92 @@ class AlertChannelDefiningMixin(object):
+                    f"Channel {token} organization {alert_receive_channel.organization.public_primary_key} is deleted"
                 )
-                raise OrganizationMovedException(alert_receive_channel.organization)
-            if alert_receive_channel.deleted_at or alert_receive_channel.organization.deleted_at:
+                cache.set(cache_key_short_term, CHANNEL_DOES_NOT_EXIST_PLACEHOLDER, self.CACHE_SHORT_TERM_TIMEOUT)
+                return None, CHANNEL_DOES_NOT_EXIST_PLACEHOLDER
+
+            # Update short term cache
+            serialized = serializers.serialize("json", [alert_receive_channel])
+            cache.set(cache_key_short_term, serialized, self.CACHE_SHORT_TERM_TIMEOUT)
+
+            # Update cached channels
+            if cache.get(self.CACHE_DB_FALLBACK_OBSOLETE_KEY) is None:
+                cache.set(self.CACHE_DB_FALLBACK_OBSOLETE_KEY, True, self.CACHE_DB_FALLBACK_REFRESH_INTERVAL)
+                self.update_alert_receive_channel_cache()
+
+        return alert_receive_channel, None
+
+    def get_alert_receive_channel_from_db(self, token: str) -> tuple[Optional[AlertReceiveChannel], bool]:
+        from apps.alerts.models import AlertReceiveChannel
+
+        try:
+            alert_receive_channel = AlertReceiveChannel.objects.get(token=token)
+            if alert_receive_channel.organization.is_moved:
+                # Lookup moved organizations every time and do not cache. Not currently compatible with fallback logic
                 logger.info(
-                    f"Channel {kwargs['alert_channel_key']} or organization {alert_receive_channel.organization.public_primary_key} is deleted"
+                    f"Channel {token} in organization {alert_receive_channel.organization.public_primary_key} is moved"
                 )
-                raise PermissionDenied(INTEGRATION_PERMISSION_DENIED_MESSAGE)
+                raise OrganizationMovedException(alert_receive_channel.organization)
 
-        del kwargs["alert_channel_key"]
+            return alert_receive_channel, True
+        except AlertReceiveChannel.DoesNotExist:
+            return None, True
+        except OperationalError:
+            logger.info("Cannot connect to database, using cache to consume alerts!")
+            return self.get_alert_receive_channel_from_db_fallback(token), False
 
-        request = args[0]
-        request.alert_receive_channel = alert_receive_channel
-        finish = perf_counter()
-        logger.info(f"AlertChannelDefiningMixin finished in {finish - start}")
-        return super(AlertChannelDefiningMixin, self).dispatch(*args, **kwargs)
+    def get_alert_receive_channel_from_db_fallback(self, token: str) -> Optional[AlertReceiveChannel]:
+        if cache.get(self.CACHE_KEY_DB_FALLBACK):
+            for obj in serializers.deserialize("json", cache.get(self.CACHE_KEY_DB_FALLBACK)):
```

Comment:
making two `cache.get` calls for the same key here. You could probably do something like:
```python3
if (objects := cache.get(self.CACHE_KEY_DB_FALLBACK)) is not None:
  for obj in serializers.deserialize("json", objects):
    ...
```